packages/hardhat-core/src/internal/hardhat-network/provider/node.ts

```diff
@@ -1040,6 +1050,40 @@ Hardhat Network's forking functionality only works with blocks from at least spu
     return true;
   }
 
+  public async dumpState(): Promise<Buffer> {
+    const state: SerializableNodeState = {
+      storage: await this._stateManager.dumpState(),
+      minTimestamp: (await this.getLatestBlock()).header.timestamp.toNumber(),
+    };
+
+    // TODO: would be way better to utilize streaming here
+    return deflatePromise(Buffer.from(JSON.stringify(state)));
+  }
+
+  public async loadState(rawState: Buffer): Promise<boolean> {
+    // TODO: would be way better to utilize streaming here
+    const deflatedState = await inflatePromise(rawState);
+
+    const state: SerializableNodeState = JSON.parse(
+      deflatedState.toString("utf8")
+    );
+
+    await this._stateManager.loadState(
+      ImmutableMap(state.storage) as ImmutableMap<string, any>
+    );
+
+    this._nextBlockTimestamp = BN.max(
+      this._nextBlockTimestamp,
+      new BN(state.minTimestamp)
+    );
+
+    // this has not been needed in testing for my tools, but it might be the right thing to do anyway.
+    // LMK if this should be called after importing new state
+    // await node._persistIrregularWorldState();
```

Comment:
I'm not sure yet what the answer is here, but maybe this comment is suggesting that we should have some tests that mix irregular states with this new persistence functionality? Note that the existing tests for that stuff are in `packages/hardhat-core/test/internal/hardhat-network/provider/modules/hardhat.ts`, exercised by things like `hardhat_setBalance`, `hardhat_setCode`, `hardhat_setNonce` and `hardhat_setStorageAt`. Note also, however, that those tests are at the RPC level, not the `HardhatNode` level, so some adaptation would need to be done, either to bring a persistence to the RPC level (which would probably be a good idea anyways) or to bring an irregularity to the `HardhatNode` level.

Reply:
sure, let me add a test for the irregular state persistance and we can figure out what to do from there based on the result :+1:

Reply:
Note to other reviewers: those tests have been added, but I'm still not yet sure what the answer is here.